### DISCUSSION TITLE
Navigation with ai deck

### DISCUSSION THREAD

**Discussion by MirelaDomiciano** [bitcraze/discussions#191] [Q&A]
hello, I was wondering if to use DroNet I have to make any changes to the crazyflie firmware.  Is it necessary to develop new firmware to receive data from the dronet and perform the movement or is this all ready?  Can modules ready in firmware be implemented using data from the deck?  Sorry if my question is a little stupid, but I'm at the beginning of my research and I still can't understand very well how I use the ai deck for autonomous navigation.

**Comment by knmcguire** [comment]
Hi!

I'm assuming that you are referring to this repository right? https://github.com/pulp-platform/pulp-dronet. 

As it goes, they send characters to the crazyflie through uart and the crazyflie turns those into hover commands that make it move, steer or stop. Currently we would advise to look at the app layer for something like that in particular. https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/

There are some videos that you can watch first too since you are at the start of your research. These are talks given by Pulp-platform about their work and explain how things work. 

- https://youtu.be/o9asYPHxEB4
- https://youtu.be/OBOBfGiq8aY

**Reply by zdh1995** [reply]
Hi,
I am confused about "they send characters to the crazyflie through uart and the crazyflie turns those into hover commands that make it move"。
I follow the guidance in [https://github.com/pulp-platform/pulp-dronet/tree/master/pulp-dronet-v2] and run the DroNet on ai deck successfully。

> ..........
> Checksum in/out Layer :	Ok
> Layer MatMul 14 ended: 
> Checksum final :	Ok
> 
> [0] : num_cycles: 7892720
> [0] : MACs: 43238144
> [0] : MAC/cycle: 26815622261777050000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000
> [0] : n. of Cores: 8

Is it meant that the dronet work on ai deck continuously ? How can I get the get the output of dronet by UART?  And I can‘t find any code about UART in the folder /application (which is the deployment C code).
Sorry if I didn't make myself clear.

**Reply by knmcguire** [reply]
Then probably the code they used for the paper is a bit different. This is unfortunately not something I can help with but you should contact pulp-platform for those specifics

**Comment by MirelaDomiciano** [comment]
Thank you for your help!! but it is necessary to use create code for programmed Dronet flight control for the app layer? The videos already watched helped me a lot, but my biggest question is a flight controller, is still not clear to me when studied the pulp-dronet. 
If necessary, this code would be a good alternative to become a base in the flight controller development? 
-> https://github.com/DP8/crazyflie-firmware/blob/master/src/deck/drivers/src/pulp_shield.c

**Reply by knmcguire** [reply]
ofcourse that is something you can do instead. Technically anything is possible since the firmware is opensource. The problem with intergrating it directly in the firmware like here, is that it is difficult to constantly update the firmware in case you want to use new features, especially those for the AIdeck. The app layer is out of tree, so you can keep your application separate from the firmware and don't have to worry about merging it with master all the time.

Practically, it all comes down at the same thing from the firmware point of view, but maintenance wise its better to make an applayer, perhaps with inspiration of the repository that you have found. This folder has some autonomous examples of the app layer that you could check out https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/demos
